### PR TITLE
Measured methods config option

### PR DESCRIPTION
**Pull request by nayeem-kamal**
# What Does This Do
Adds the ability to choose traced methods to measure through the config variables.
# Motivation
Customers want the ability to measure spans by choice rather than only have service entry spans measured.
# Additional Notes
The format for the config variable input is the same as defined for dd.trace.methods as shown here: https://docs.datadoghq.com/tracing/trace_collection/custom_instrumentation/java/

### CODE REVIEW

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+
+import datadog.trace.api.InstrumenterConfig;
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+public class MeasuredMethodFilter {
+  public static final MeasuredMethodFilter FILTER = new MeasuredMethodFilter();
+  private static Map<String, Set<String>> methodsToMeasure;
```
Comment (mcculls) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java:11]: this should be `private final Map<String, Set<String>> methodsToMeasure;` given it's initialized in the constructor and accessed by the `filter` method

... but it might just be simpler to make this a static utility class, ie:
* make `methodsToMeasure` `static final` and initialize it with a `static {}` block instead of a constructor
* make the `filter` method static

I'd also suggest adding a `static final boolean` flag which you can use to short-circuit the `filter` check, because the JIT can then optimize away the more expensive set membership test when there are no measured methods.

File: internal-api/src/main/java/datadog/trace/api/InstrumenterConfig.java [hunk 2]
```diff
@@ -383,6 +390,8 @@ public String toString() {
         + ", traceMethods='"
         + traceMethods
         + '\''
+        + measureMethods
```
Comment (mcculls) [inline internal-api/src/main/java/datadog/trace/api/InstrumenterConfig.java:393]: looks like this is missing a line before it, such as `+ ", measureMethods='"`

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/TraceAnnotationConfigParser.java [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+  }
+
+  private static boolean isIllegalMethodName(String string) {
+    return !string.equals("*") && hasIllegalCharacters(string);
+  }
+
+  @SuppressForbidden
+  public static Map<String, Set<String>> parse(String configString) {
+    Map<String, Set<String>> classMethodsToTrace;
+    if (configString.isEmpty() || configString == null) {
```
Comment (amarziali) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/TraceAnnotationConfigParser.java:47]: the null test should be before the .isEmpty to avoid a NPE

**Comment by mcculls** [comment]
@nayeem-kamal FYI I had to merge the muzzle fix from this PR into master to fix a related issue, so you'll need to merge in the latest changes from master and resolve the conflict before this can go in

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+  public MeasuredMethodFilter() {
+    if (InstrumenterConfig.get().getMeasureMethods().isEmpty()
+        || InstrumenterConfig.get().getMeasureMethods() == null) {
```
[the remark this answers is not in the document]
Comment (nayeem-kamal) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java]: Thanks for the review @amarziali. This should be fixed now.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java (hunk 1 above)
Comment (nayeem-kamal) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java]: Thanks Stuart! Yeah the static utility class seemed like the way to go so I made that change.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/TraceAnnotationConfigParser.java (hunk 3 above)
Comment (amarziali) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/TraceAnnotationConfigParser.java]: why this is solved? Either there is no need to test for null (so the second member can be removed) either you should inverse the or condition `configString == null` before

Comment (nayeem-kamal) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/TraceAnnotationConfigParser.java]: @amarziali I believe the code snippet shown above is outdated. In my recent commit, I flipped the condition to check for null before checking isEmpty.

Comment (nayeem-kamal) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/TraceAnnotationConfigParser.java]: Sorry for the confusion!

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java (hunk 4 above)
[the remark this answers is not in the document]
Comment (amarziali) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java]: yes you're right. I had to mess with github :D

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+      filterIsEmpty = methodsToMeasure.isEmpty();
+    }
+  }
+
+  public static boolean filter(Method method) {
+    if (!filterIsEmpty) {
+      String clazz = method.getDeclaringClass().getName();
+      return methodsToMeasure.containsKey(clazz)
+          && (methodsToMeasure.get(clazz).contains(method.getName())
+              || methodsToMeasure.get(clazz).contains("*"));
```
Comment (mcculls) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java:29]: You can avoid repeated hash lookups here by only fetching the value once:
```
      Set<String> methods = methodsToMeasure.get(clazz);
      return methods != null && (methods.contains(method.getName()) || methods.contains("*"));
```

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+import java.util.Map;
+import java.util.Set;
+
+public class MeasuredMethodFilter {
+  private static final Map<String, Set<String>> methodsToMeasure;
+  private static final boolean filterIsEmpty;
+
+  static {
+    String configString = InstrumenterConfig.get().getMeasureMethods();
+    if (configString == null || configString.isEmpty()) {
```
Comment (ygree) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java:15]: This extra check can be removed completely because it's already in `TraceAnnotationConfigParser.parse`

File: dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java [hunk 7]
```diff
@@ -19,6 +19,7 @@ public final class TraceInstrumentationConfig {
   public static final String TRACE_EXECUTORS_ALL = "trace.executors.all";
   public static final String TRACE_EXECUTORS = "trace.executors";
   public static final String TRACE_METHODS = "trace.methods";
+  public static final String MEASURE_METHODS = "measure.methods";
```
Comment (ygree) [inline dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java:22]: Would be nice to have this setting format documented somewhere other than in tests. Maybe add a comment?

**Review by ygree — COMMENTED** [review]
LGTM. Only a couple of small suggestions.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/TraceAnnotationConfigParser.java [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+  private static boolean isIllegalClassName(String string) {
+    return hasIllegalCharacters(string);
+  }
+
+  private static boolean isIllegalMethodName(String string) {
+    return !string.equals("*") && hasIllegalCharacters(string);
+  }
+
+  @SuppressForbidden
+  public static Map<String, Set<String>> parse(String configString) {
```
Comment (ygree) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/TraceAnnotationConfigParser.java:45]: Nice reuse.

File: dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+import java.lang.reflect.Method;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Set;
+
+public class MeasuredMethodFilter {
+  private static final Map<String, Set<String>> methodsToMeasure;
+  private static final boolean filterIsEmpty;
+
+  static {
```
Comment (ygree) [inline dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java:13]: Does it really need to be static?

Comment (mcculls) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java:13]: The `filterIsEmpty` boolean should be `static final` somewhere, so the JVM can optimize away the branch when there's no measured methods (https://shipilev.net/jvm/anatomy-quarks/15-just-in-time-constants/)

Comment (ygree) [reply dd-java-agent/agent-bootstrap/src/main/java/datadog/trace/bootstrap/instrumentation/traceannotation/MeasuredMethodFilter.java:12]: Thanks, @mcculls!

File: dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java (hunk 7 above)
Comment (ygree) [reply dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java:26]: Thanks, @nayeem-kamal!

**Review by ygree — APPROVED** [review]
Looks good to me!

File: dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java [hunk 10]
```diff
@@ -19,6 +19,11 @@ public final class TraceInstrumentationConfig {
   public static final String TRACE_EXECUTORS_ALL = "trace.executors.all";
   public static final String TRACE_EXECUTORS = "trace.executors";
   public static final String TRACE_METHODS = "trace.methods";
+  /*
+  format for measure.methods is the same as for trace.methods:
+  https://docs.datadoghq.com/tracing/trace_collection/custom_instrumentation/java/
+   */
+  public static final String MEASURE_METHODS = "measure.methods";
```
Comment (TheSench) [inline dd-trace-api/src/main/java/datadog/trace/api/config/TraceInstrumentationConfig.java:26]: It would be nice to have this setting documented in the custom instrumentation documentation, similar to the `dd.trace.methods` configuration that you [linked to](https://docs.datadoghq.com/tracing/trace_collection/custom_instrumentation/java/) in your additional notes section.

It is currently not discoverable by customers unless they stumble across this PR or dig into the agent code themselves.
